auto_ptr is deprecated in C++11

auto_ptr is [deprecated in C++11](https://en.cppreference.com/w/cpp/memory/auto_ptr) and should be replaced with unique_ptr according to [gotw #89](https://herbsutter.com/2013/05/29/gotw-89-solution-smart-pointers/).

Excerpts from build output of GCC 12.1.0 on MSys2 Mingw64 shell on Windows 10:
```
src/MapScriptUSL.cpp: In member function 'bool MapScriptUSL::compileCode(const std::string&)':
src/MapScriptUSL.cpp:181:41: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
  181 |                                         auto_ptr<ifstream> file(Toolkit::getFileManager()->openIFStream(fullFileName));
      |                                         ^~~~~~~~
In file included from D:/msys64/mingw64/include/c++/12.1.0/memory:76,
                 from D:/msys64/mingw64/include/boost/smart_ptr/detail/sp_counted_impl.hpp:35,
                 from D:/msys64/mingw64/include/boost/smart_ptr/detail/shared_count.hpp:27,
                 from D:/msys64/mingw64/include/boost/smart_ptr/shared_ptr.hpp:17,
                 from D:/msys64/mingw64/include/boost/shared_ptr.hpp:17,
                 from src/Team.h:33,
                 from src/Map.h:31,
                 from src/Game.h:27,
                 from src/GameGUI.h:28,
                 from src/MapScriptUSL.cpp:25:
D:/msys64/mingw64/include/c++/12.1.0/bits/unique_ptr.h:64:28: note: declared here
   64 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~

libusl/src/parser.cpp: In member function 'ExpressionNode* Parser::declaration2(const Position&)':
libusl/src/parser.cpp:86:17: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
   86 |                 auto_ptr<PatternNode> argument(pattern());
      |                 ^~~~~~~~
In file included from D:/msys64/mingw64/include/c++/12.1.0/memory:76,
                 from D:/msys64/mingw64/include/boost/function/function_base.hpp:16,
                 from D:/msys64/mingw64/include/boost/function/detail/prologue.hpp:17,
                 from D:/msys64/mingw64/include/boost/function.hpp:30,
                 from libusl/src/native.h:10:
D:/msys64/mingw64/include/c++/12.1.0/bits/unique_ptr.h:64:28: note: declared here
   64 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~
libusl/src/parser.cpp: In member function 'PatternNode* Parser::pattern()':
libusl/src/parser.cpp:107:25: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
  107 |                         auto_ptr<TuplePatternNode> tuple(new TuplePatternNode(position));
      |                         ^~~~~~~~
D:/msys64/mingw64/include/c++/12.1.0/bits/unique_ptr.h:64:28: note: declared here
   64 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~
libusl/src/parser.cpp: In member function 'ExpressionNode* Parser::methodCompositionExpression(ExpressionNode*)':
libusl/src/parser.cpp:192:9: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
  192 |         auto_ptr<ExpressionNode> node(pathExpression(first));
      |         ^~~~~~~~
D:/msys64/mingw64/include/c++/12.1.0/bits/unique_ptr.h:64:28: note: declared here
   64 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~
libusl/src/parser.cpp: In member function 'ExpressionNode* Parser::pathExpression(ExpressionNode*)':
libusl/src/parser.cpp:214:9: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
  214 |         auto_ptr<ExpressionNode> node(first);
      |         ^~~~~~~~
D:/msys64/mingw64/include/c++/12.1.0/bits/unique_ptr.h:64:28: note: declared here
   64 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~
libusl/src/parser.cpp: In member function 'ExpressionNode* Parser::simpleExpression()':
libusl/src/parser.cpp:269:25: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
  269 |                         auto_ptr<RecordBlock> block(new RecordBlock(position));
      |                         ^~~~~~~~
D:/msys64/mingw64/include/c++/12.1.0/bits/unique_ptr.h:64:28: note: declared here
   64 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~
libusl/src/parser.cpp:289:25: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
  289 |                         auto_ptr<ExecutionBlock> block(new ExecutionBlock(position));
      |                         ^~~~~~~~
D:/msys64/mingw64/include/c++/12.1.0/bits/unique_ptr.h:64:28: note: declared here
   64 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~
libusl/src/parser.cpp:297:25: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
  297 |                         auto_ptr<RecordBlock> block(new RecordBlock(position));
      |                         ^~~~~~~~
D:/msys64/mingw64/include/c++/12.1.0/bits/unique_ptr.h:64:28: note: declared here
   64 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~
libusl/src/parser.cpp:305:25: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
  305 |                         auto_ptr<PatternNode> arg(pattern());
      |                         ^~~~~~~~
D:/msys64/mingw64/include/c++/12.1.0/bits/unique_ptr.h:64:28: note: declared here
   64 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~
libusl/src/usl.cpp: In member function 'virtual void Load::execute(Thread*)':
libusl/src/usl.cpp:64:17: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
   64 |                 auto_ptr<ifstream> stream(usl->openFile(filename));
      |                 ^~~~~~~~
In file included from D:/msys64/mingw64/include/c++/12.1.0/memory:76,
                 from D:/msys64/mingw64/include/boost/function/function_base.hpp:16,
                 from D:/msys64/mingw64/include/boost/function/detail/prologue.hpp:17,
                 from D:/msys64/mingw64/include/boost/function.hpp:30,
                 from libusl/src/native.h:10:
D:/msys64/mingw64/include/c++/12.1.0/bits/unique_ptr.h:64:28: note: declared here
   64 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~
```

--- a/libusl/src/parser.cpp
+++ b/libusl/src/parser.cpp
@@ -83,7 +83,7 @@ ExpressionNode* Parser::declaration2(const Position& position) {
 		next();
 		break;
 	default:
-		auto_ptr<PatternNode> argument(pattern());
+		unique_ptr<PatternNode> argument(pattern());
 		ExpressionNode* body = declaration2(position);
 		return new FunNode(position, argument.release(), body);
 	}
@@ -104,7 +104,7 @@ PatternNode* Parser::pattern()
 		}
 		else
 		{
-			auto_ptr<TuplePatternNode> tuple(new TuplePatternNode(position));
+			unique_ptr<TuplePatternNode> tuple(new TuplePatternNode(position));
 			tuple->members.push_back(pattern());
 			while (tokenType() == COMMA)
 			{
@@ -189,7 +189,7 @@ ExpressionNode* Parser::prefixedExpression()
 
 ExpressionNode* Parser::methodCompositionExpression(ExpressionNode* first)
 {
-	auto_ptr<ExpressionNode> node(pathExpression(first));
+	unique_ptr<ExpressionNode> node(pathExpression(first));
 	while (true)
 	{
 		const Position position(token.position);
@@ -211,7 +211,7 @@ ExpressionNode* Parser::methodCompositionExpression(ExpressionNode* first)
 
 ExpressionNode* Parser::pathExpression(ExpressionNode* first)
 {
-	auto_ptr<ExpressionNode> node(first);
+	unique_ptr<ExpressionNode> node(first);
 	while (true)
 	{
 		const Position& position = token.position;
@@ -266,7 +266,7 @@ ExpressionNode* Parser::simpleExpression()
 	case LPAR:
 		{
 			next();
-			auto_ptr<RecordBlock> block(new RecordBlock(position));
+			unique_ptr<RecordBlock> block(new RecordBlock(position));
 			expressions(block.get());
 			accept(RPAR);
 			RecordBlock* record = block.get();
@@ -286,7 +286,7 @@ ExpressionNode* Parser::simpleExpression()
 	case LBRACE:
 		{
 			next();
-			auto_ptr<ExecutionBlock> block(new ExecutionBlock(position));
+			unique_ptr<ExecutionBlock> block(new ExecutionBlock(position));
 			statements(block.get());
 			accept(RBRACE);
 			return block.release();
@@ -294,7 +294,7 @@ ExpressionNode* Parser::simpleExpression()
 	case LBRACK:
 		{
 			next();
-			auto_ptr<RecordBlock> block(new RecordBlock(position));
+			unique_ptr<RecordBlock> block(new RecordBlock(position));
 			statements(block.get());
 			accept(RBRACK);
 			return block.release();
@@ -302,7 +302,7 @@ ExpressionNode* Parser::simpleExpression()
 	case FUN:
 		{
 			next();
-			auto_ptr<PatternNode> arg(pattern());
+			unique_ptr<PatternNode> arg(pattern());
 			accept(ARROW);
 			ExpressionNode* body = expression();
 			return new FunNode(position, arg.release(), body);

--- a/libusl/src/usl.cpp
+++ b/libusl/src/usl.cpp
@@ -61,7 +61,7 @@ struct Load: NativeCode
 		
 		Usl* usl = thread->usl;
 		
-		auto_ptr<ifstream> stream(usl->openFile(filename));
+		unique_ptr<ifstream> stream(usl->openFile(filename));
 		Scope* scope = usl->compile(filename, *stream);
 		if (frame.nextInstr >= frame.thunk->thunkPrototype()->body.size())
 			thread->frames.pop_back(); // tail-call optimisation

--- a/src/MapScriptUSL.cpp
+++ b/src/MapScriptUSL.cpp
@@ -178,7 +178,7 @@ bool MapScriptUSL::compileCode(const std::string& code)
 				do {
 					fileName = Toolkit::getFileManager()->getNextDirectoryEntry();
 					std::string fullFileName = string(*dir) + DIR_SEPARATOR + fileName;
-					auto_ptr<ifstream> file(Toolkit::getFileManager()->openIFStream(fullFileName));
+					unique_ptr<ifstream> file(Toolkit::getFileManager()->openIFStream(fullFileName));
 					if (file.get())
 					{
 						#ifdef DEBUG_USL

